zproject/backends.py

```diff
@@ -182,36 +197,20 @@ def authenticate(self,
     def _common_authenticate(self, *args, **kwargs):
         # type: (*Any, **Any) -> Optional[UserProfile]
         return_data = kwargs.get('return_data', {})
+        realm = kwargs.get("realm")
+        if realm is None:
+            return None
+        if not auth_enabled_helper([self.auth_backend_name], realm):
+            return_data["auth_backend_disabled"] = True
+            return None
 
         email_address = self.get_email_address(*args, **kwargs)
         if not email_address:
             return_data['invalid_email'] = True
             return None
 
-        try:
-            user_profile = get_user_profile_by_email(email_address)
-        except UserProfile.DoesNotExist:
-            return_data["valid_attestation"] = True
-            return None
-
-        if not user_profile.is_active:
-            return_data["inactive_user"] = True
-            return None
-
-        if user_profile.realm.deactivated:
-            return_data["inactive_realm"] = True
-            return None
-
-        if not user_matches_subdomain(kwargs.get("realm_subdomain"),
-                                      user_profile):
-            return_data["invalid_subdomain"] = True
-            return None
-
-        if not auth_enabled_helper([self.auth_backend_name], user_profile.realm):
-            return_data["auth_backend_disabled"] = True
-            return None
-
-        return user_profile
+        return_data["valid_attestation"] = True
```

Comment:
Yes, it'll break things to delete this.  This is an artifact of how Python Social Auth works.  It doesn't actually call `authenticate()` directly; instead, it calls `do_auth`, which in turn calls `authenticate` (and then `process_do_auth`).  So we actually do know it's the case, it's just confusing.

There's more I want to do to refactor this backend to be more readable (this isn't the only confusing thing in this backend); my plan is to leave that for a follow-up project since it isn't directly related.

Reply:
Your latest branch separates out this behavior change as its own commit -- that's helpful, thanks!

Reply:
Hmm, I'm a little confused, though -- one of your commit messages now says `valid_attestation` doesn't seem to be used by the caller. What does it break -- is it only too-specific tests, or something else?

Reply:
I think you must have been looking at an older version before, which is perhaps why I'm confused.  I made the change to split that into its own commit at 3:40PM today... Maybe the problem is this was the 3rd-to-last-commit (covered by "not done yet") yesterday, I guess?

Anyway, what did need to change is that later commit message (the original commit for the GitHub series that was later split into 5 commits) was a mistake; I've removed that text.

Reply:
(It's not too-specific tests; it's used in `api_fetch_api_key` to give a specific error message for mobile auth)

Reply:
I was referring to this commit message, with a commit date from you of 6:35PM today:

```
commit 072d07cc89a6b0bc19ce840c670d5a7f21522a47
Author:     Tim Abbott <tabbott@zulipchat.com>
AuthorDate: Tue Nov 21 14:51:22 2017 -0800
Commit:     Tim Abbott <tabbott@zulipchat.com>
CommitDate: Tue Nov 21 18:35:05 2017 -0800

    auth: Convert SocialAuthMixin to use new helper.
    
    The `valid_attestation` variable doesn't seem to be used by the
    caller, but if it were, it should be set before we check if the user
    account is valid.
```

Oh I see, you mean just the split happened earlier. Yeah, it looks like I first fetched this branch at 1319 today, then next at 1826. Reminder that when you push an update to a branch, that's nearly invisible on GitHub, and a comment is very handy if you want a reviewer to see the update. :-)

Reply:
That was this comment at 4:21PM: https://github.com/zulip/zulip/pull/7472#issuecomment-346203769

Reply:
Hmm, I see. That may be another case of comments racing, then -- I think I didn't see that comment until sometime after my second round of comments (including the start of this subthread.) Not sure there's a solution here other than more vigorously polling the thread when in the middle of review.